### ISSUE TITLE
B::Xref seems to be broken

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6069]
Migrated from [rt.perl.org#18306](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=18306) (status was 'resolved')

Searchable as RT18306$

**Comment by p5pRT** [comment]
### From at@localhost.localdomain

#### Created by at@turbinal.org



$ perl \-MO=Xref /usr/lib/perl5/find\.pl | grep ^File
File
File /usr/lib/perl5/Carp\.pm
File /usr/lib/perl5/Exporter\.pm
File /usr/lib/perl5/File/Basename\.pm
File /usr/lib/perl5/File/Find\.pm
File /usr/lib/perl5/File/Spec/Unix\.pm
File /usr/lib/perl5/Scalar/Util\.pm
File /usr/lib/perl5/base\.pm
File /usr/lib/perl5/find\.pl
File /usr/lib/perl5/i386\-linux/Cwd\.pm
File /usr/lib/perl5/i386\-linux/DynaLoader\.pm
File /usr/lib/perl5/i386\-linux/XSLoader\.pm
File /usr/lib/perl5/i386\-linux/re\.pm
/usr/lib/perl5/find\.pl syntax OK
File @&#8203;EXPORT
File Exporter
File PPp����\*
File \`\`p����\*
File ��&��\*��������p��
$

These last 5 files don't seem to be files at all\. This is also very suspicious
that these filenames contain non\-latin characters\, this leads me to the point
that there's a memory\-related error somewhere inside of B&#8203;::Xref\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by at at Tue Nov  5 19:22:33 MSK 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18-alt6master-up, archname=i386-linux-thread-multi
    uname='linux homestead.turbinal.org 2.4.18-alt6master-up #1 tue apr 16 14:50:56 msd 2002 i686 unknown unknown gnulinux '
    config_args='-des -Darchname=i386-linux -Dd_dosuid -Ud_csh -Duseshrplib -Dlibperl=libperl.so.5.8 -Doptimize=-pipe -Wall -O2 -fexpensive-optimizations -march=i686 -Dinstallprefix=/usr -Dprefix=/usr -Dprivlib=/usr/lib/perl5 -Darchlib=/usr/lib/perl5/i386-linux -Dvendorprefix=/usr -Dvendorlib=/usr/lib/perl5/vendor_perl -Dvendorarch=/usr/lib/perl5/vendor_perl/i386-linux -Dsiteprefix=/usr/local -Dsitelib=/usr/local/lib/perl5/site_perl/5.8.0 -Dsitearch=/usr/local/lib/perl5/site_perl/5.8.0/i386-linux -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dcf_by=ALT Linux Team -Dmyhostname=localhost -Dperladmin=root@localhost -Dinc_version_list=5.6.1 5.6.0  -Dpager=/usr/bin/less -isR -Di_shadow -Di_syslog -Dusethreads -Duseithreads -Duselargefiles -Ui_db -Di_gdbm -Di_ndbm -Di_sdbm -Ui_odbm'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-pipe -Wall -O2 -fexpensive-optimizations -march=i686',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.1 20020822 (ALT Linux, build 3.2.1-alt0.3)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.6.so, so=so, useshrplib=true, libperl=libperl.so.5.8
    gnulibc_version='2.2.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/i386-linux
    /usr/lib/perl5
    /usr/local/lib/perl5/site_perl/5.8.0/i386-linux
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/i386-linux
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.0:
    HOME=/home/at
    LANG=ru_RU.CP1251
    LANGUAGE=C
    LC_COLLATE=ru_RU.CP1251
    LC_CTYPE=ru_RU.CP1251
    LC_MESSAGES=C
    LC_MONETARY=ru_RU.CP1251
    LC_NUMERIC=ru_RU.CP1251
    LC_TIME=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/bin:/usr/bin:/usr/local/bin:/usr/X11R6/bin:/home/at/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=zsh

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

at@&#8203;localhost\.localdomain \(via RT\) wrote&#8203;:
> 
> $ perl \-MO=Xref /usr/lib/perl5/find\.pl | grep ^File
> File
> File /usr/lib/perl5/Carp\.pm
> File /usr/lib/perl5/Exporter\.pm
> File /usr/lib/perl5/File/Basename\.pm
> File /usr/lib/perl5/File/Find\.pm
> File /usr/lib/perl5/File/Spec/Unix\.pm
> File /usr/lib/perl5/Scalar/Util\.pm
> File /usr/lib/perl5/base\.pm
> File /usr/lib/perl5/find\.pl
> File /usr/lib/perl5/i386\-linux/Cwd\.pm
> File /usr/lib/perl5/i386\-linux/DynaLoader\.pm
> File /usr/lib/perl5/i386\-linux/XSLoader\.pm
> File /usr/lib/perl5/i386\-linux/re\.pm
> /usr/lib/perl5/find\.pl syntax OK
> File @&#8203;EXPORT
> File Exporter
> File PPpÃÂ´\*
> File \`\`pÃÂ´\*
> File Â&Ã\*Ã¾Ã¿Ã¿Ã¿pÂ
> $
> 
> These last 5 files don't seem to be files at all\. This is also very suspicious
> that these filenames contain non\-latin characters\, this leads me to the point
> that there's a memory\-related error somewhere inside of B&#8203;::Xref\.

I can reproduce this\, but only with \-Dusethreads\.
\(Unfortunately valgrind doesn't help in this case\.\)

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

\[at@&#8203;localhost\.localdomain \- Sun Nov 10 00&#8203;:17&#8203;:08 2002\]&#8203;:

> 
\[\.\.\.\]
> File PPpÃ´\*
> File \`\`pÃ´\*
> File &Ô\*þÿÿÿp

My Bug \#6671 looks similar \(file name contains garbage\)&#8203;:

 \<http&#8203;://bugs6\.perl\.org/rt2/Ticket/Display\.html?id=6671>


Steffen

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Sun Nov 10 00&#8203;:17&#8203;:08 2002\, at@&#8203;localhost\.localdomain wrote&#8203;:
> 
> $ perl \-MO=Xref /usr/lib/perl5/find\.pl | grep ^File
> File
> File /usr/lib/perl5/Carp\.pm
> File /usr/lib/perl5/Exporter\.pm
> File /usr/lib/perl5/File/Basename\.pm
> File /usr/lib/perl5/File/Find\.pm
> File /usr/lib/perl5/File/Spec/Unix\.pm
> File /usr/lib/perl5/Scalar/Util\.pm
> File /usr/lib/perl5/base\.pm
> File /usr/lib/perl5/find\.pl
> File /usr/lib/perl5/i386\-linux/Cwd\.pm
> File /usr/lib/perl5/i386\-linux/DynaLoader\.pm
> File /usr/lib/perl5/i386\-linux/XSLoader\.pm
> File /usr/lib/perl5/i386\-linux/re\.pm
> /usr/lib/perl5/find\.pl syntax OK
> File @&#8203;EXPORT
> File Exporter
> File PPpô\*
> File \`\`pô\*
> File �&�\*����p�
> $
> 
> These last 5 files don't seem to be files at all\. This is also very
>    suspicious
> that these filenames contain non\-latin characters\, this leads me to
>    the point
> that there's a memory\-related error somewhere inside of B&#8203;::Xref\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

perl\-5\.8\.8 \-MO=Xref lib/find\.pl | grep ^File 
File
File DESTROY
File hPnd&#8203;::prune
File import
File lib/Cwd\.pm
File lib/File/Basename\.pm
File lib/File/Find\.pm
File lib/File/Spec/Unix\.pm
File lib/Scalar/Util\.pm
File lib/XSLoader\.pm
File lib/find\.pl
File lib/re\.pm
lib/find\.pl syntax OK
File È­
File ¥5Y
File Èäääe\.pm


perl\-5\.10\.0 \-MO=Xref find\.pl | grep ^File
Can't locate object method "NAME" via package "B&#8203;::SPECIAL" at lib/B/
Xref\.pm line 162\.
CHECK failed\-\-call queue aborted\.

perl\-blead \-MO=Xref find\.pl | grep ^File
Can't locate object method "NAME" via package "B&#8203;::SPECIAL" at lib/B/
Xref\.pm line 162\.
CHECK failed\-\-call queue aborted\.

All compiled with&#8203;:

\./Configure \-Dusethreads \-Duseithreads \-Dusedevel=y \-des

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

I wanted to test run the "find" script\. As "animator" mentioned it
couldn't be run because B&#8203;::Xref tries to call a method "NAME" on a
B&#8203;::SPECIAL object\.

So I patched B&#8203;::Xref and run the script\. The problem the OP had seems to
be fixed\.\.\.

The output&#8203;:

\./perl \-Ilib \-MO=Xref lib/find\.pl | grep ^File
File lib/Cwd\.pm
File lib/File/Basename\.pm
File lib/File/Find\.pm
File lib/File/Spec/Unix\.pm
File lib/List/Util\.pm
File lib/Scalar/Util\.pm
File lib/XSLoader\.pm
File lib/find\.pl
lib/find\.pl syntax OK
File lib/re\.pm

\./perl \-Ilib \-V
Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.21\-xen\-vs2\.2\.0\.4\,
archname=x86\_64\-linux\-thread\-multi
&nbsp;    uname='linux www\-devel\-rbaecker 2\.6\.21\-xen\-vs2\.2\.0\.4 \#1 smp fri apr
18 19&#8203;:03&#8203;:32 cest 2008 x86\_64 intel\(r\) xeon\(r\) cpu e5430 @&#8203; 2\.66ghz
genuineintel gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duseithreads \-Dusedevel=y \-des'
&nbsp;    hint=previous\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
\.\.\.\.


Cheers\,
Renee


---
renee\.baecker@&#8203;smart\-websolutions\.de

Smart\-Websolutions André Windolph und Renée Bäcker GbR
Maria\-Montessori\-Str\. 13
64584 Biebesheim

USt\.\-ID&#8203;: DE 228 935 695

Telefon&#8203;: 01803 \- 278 25 51 98 \(9 cent/Minute\)
E\-Mail&#8203;: info@&#8203;smart\-websolutions\.de

Perl\-Magazin&#8203;: http&#8203;://foo\-magazin\.de
Perl\-Nachrichten&#8203;: http&#8203;://perl\-nachrichten\.de

**Comment by p5pRT** [comment]
### From renee.baecker@smart-websolutions.de

<details><summary>Xref.patch</summary>

```diff
--- /home/rbaecker/perl510/perl-5.10.0/lib/B/Xref.pm	2007-12-18 11:47:07.000000000 +0100
+++ Xref.pm	2008-07-09 15:07:14.000000000 +0200
@@ -157,6 +157,7 @@
 	for ($ix = 1; $ix < @vallist; $ix++) {
 	    my $valsv = $vallist[$ix];
 	    next unless class($valsv) eq "GV";
+            next if class($valsv->STASH) eq 'SPECIAL';
 	    # these pad GVs don't have corresponding names, so same @pad
 	    # array can be used without collisions
 	    $pad[$ix] = [$valsv->STASH->NAME, "*", $valsv->NAME];

```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

On Wed Jul 09 06&#8203;:17&#8203;:39 2008\, renee\.baecker@&#8203;smart\-websolutions\.de wrote&#8203;:
> I wanted to test run the "find" script\. As "animator" mentioned it
> couldn't be run because B&#8203;::Xref tries to call a method "NAME" on a
> B&#8203;::SPECIAL object\.
> 
> So I patched B&#8203;::Xref and run the script\. The problem the OP had seems to
> be fixed\.\.\.
> 
> The output&#8203;:
> 
> \./perl \-Ilib \-MO=Xref lib/find\.pl | grep ^File
> File lib/Cwd\.pm
> File lib/File/Basename\.pm
> File lib/File/Find\.pm
> File lib/File/Spec/Unix\.pm
> File lib/List/Util\.pm
> File lib/Scalar/Util\.pm
> File lib/XSLoader\.pm
> File lib/find\.pl
> lib/find\.pl syntax OK
> File lib/re\.pm
> 

This patch was applied as change \#34772\.  Thanks\!

Steve

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
